graph/src/components/ethereum/network.rs

```diff
@@ -0,0 +1,227 @@
+}
+
+impl PartialOrd for NodeCapabilities {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl fmt::Display for NodeCapabilities {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            NodeCapabilities {
+                archive: true,
+                traces: true,
+            } => write!(f, "archive, trace"),
+            NodeCapabilities {
+                archive: false,
+                traces: true,
+            } => write!(f, "full, trace"),
+            NodeCapabilities {
+                archive: false,
+                traces: false,
+            } => write!(f, "full"),
+            NodeCapabilities {
+                archive: true,
+                traces: false,
+            } => write!(f, "archive"),
+        }
+    }
+}
+
+impl_slog_value!(NodeCapabilities, "{:?}");
+
+#[derive(Clone)]
+pub struct EthereumNetworkAdapter {
+    pub capabilities: NodeCapabilities,
+    adapter: Arc<dyn EthereumAdapter>,
+}
+
+#[derive(Clone)]
+pub struct EthereumNetworkAdapters {
+    pub adapters: Vec<EthereumNetworkAdapter>,
+}
+
+impl EthereumNetworkAdapters {
+    pub fn cheapest_with(
+        &self,
+        required_capabilities: &NodeCapabilities,
+    ) -> Result<&Arc<dyn EthereumAdapter>, Error> {
+        let sufficient_adapters: Vec<&EthereumNetworkAdapter> = self
+            .adapters
+            .iter()
+            .filter(|adapter| &adapter.capabilities >= required_capabilities)
+            .collect();
+        if sufficient_adapters.is_empty() {
+            return Err(format_err!(
+                "A matching Ethereum network with {:?} was not found.",
+                required_capabilities
+            ));
+        }
+        // TODO: Randomly choose between capable nodes
+        Ok(&sufficient_adapters.iter().next().unwrap().adapter)
+    }
+
+    pub fn cheapest(&self) -> Option<&Arc<dyn EthereumAdapter>> {
+        self.adapters
+            .iter()
+            .next()
+            .map(|ethereum_network_adapter| &ethereum_network_adapter.adapter)
+    }
+}
+
+#[derive(Clone)]
+pub struct EthereumNetworks {
+    pub networks: HashMap<String, EthereumNetworkAdapters>,
+}
+
+impl EthereumNetworks {
+    pub fn new() -> EthereumNetworks {
+        EthereumNetworks {
+            networks: HashMap::new(),
+        }
+    }
+
+    pub fn insert_or_update(
+        &mut self,
+        name: String,
+        capabilities: NodeCapabilities,
+        adapter: Arc<dyn EthereumAdapter>,
+    ) {
+        let network_adapters = self
+            .networks
+            .entry(name)
+            .or_insert(EthereumNetworkAdapters {
+                adapters: vec![EthereumNetworkAdapter {
+                    capabilities,
+                    adapter: adapter.clone(),
+                }],
+            });
+        network_adapters.adapters.push(EthereumNetworkAdapter {
+            capabilities,
+            adapter: adapter.clone(),
```

Comment:
In case there was no entry for the network yet, this is adding the adapter twice. You could do this:
```rust
        let network_adapters = self
            .networks
            .entry(name)
            .or_insert(EthereumNetworkAdapters {
                adapters: vec![],
            });

        network_adapters.adapters.push(EthereumNetworkAdapter {
            capabilities,
            adapter: adapter.clone(),
        });
```